packages/babel-core/src/parser/index.ts

```diff
@@ -49,6 +49,13 @@ export default function* parser(
       // err.code will be changed to BABEL_PARSE_ERROR later.
     }
 
+    const startLine = parserOpts?.startLine;
+    const startColumn = parserOpts?.startColumn;
+
+    if (startLine != null || startColumn != null) {
+      code = "\n".repeat(startLine - 1) + " ".repeat(startColumn) + code;
+    }
+
     const { loc, missingPlugin } = err;
     if (loc) {
```

Comment:
Can we apply the offset to the error location here? So that we don't have to copy the input code, which could be huge.

Reply:
That would display a start line and start column that would be confusing for the user.

Reply:
Understood, in that case can we pass the startLine / startColumn info to babel-code-frame and let it adjust its line markers? It think this approach should be more efficient than padding the input because the startLine / startColumn can be arbitrarily large.

Reply:
As far as I remember, the reason we recently also introduced .startIndex is because the workaround of adding all the whitespace at the beginning can be very expensive on large files.

However, this shouldn't be too bad, because we are only adding one characcter per line (\n) rather than re-filling all the previous characters. And the `+ code` likely doesn't actually copy the string, since V8 represents strings as trees/ropes of smaller strings? Unless then we do something that causes the string to be "flattened".

Also, this branch is only running in the error case, right?

Reply:
Yes, I'm fine with both approaches.

Reply:
@liuxingbaoyu What about adding `startLine`/`startColumn` to `code-frame` with a PR for the upcoming minor release? Or maybe just `startLine`, since `startColumn`'s behavior would get weird when there are multiple lines.

Reply:
Sounds good, I'll get it done.